### ISSUE TITLE
High CPU Usage when sim is paused

### ISSUE THREAD

**Issue by samreid** [phetsims/circuit-construction-kit-common#588]
Reported in https://github.com/phetsims/circuit-construction-kit-common/issues/586

MD said:

> I notice that my PC fan comes on when I run this sim. Sim seems to need a lot or CPU time and memory to run -- about 50% CPU according to task manager. What is especially puzzling is that the CPU usage does not decrease significantly when I pause the sim.

We are now running the model when the sim is paused as part of https://github.com/phetsims/circuit-construction-kit-common/issues/572.  That issue is still open and requires testing, but this issue in particular asks whether we can use that same strategy but make it less CPU intensive.

**Comment by samreid** [comment]
Most of the CPU expense is from this code in ChargeNode:

```js
      // Apply the transform
      this.matrix = NODE_MATRIX;
```

The electron moves by a tiny bit (1E-11) but this triggers scenery SVG attribute setting which takes around 10% CPU:

![image](https://user-images.githubusercontent.com/679486/92721353-3497f400-f323-11ea-92f4-fa61f4f507b6.png)

**Comment by samreid** [comment]
By checking if the matrix has changed, we can avoid that `setAttribute` and `updateLayerTree`, saving >=23% when the simulation is paused.  However, this comes at an expense of 0.6% (whether paused or not) to the simulation to compare these matrices:

![image](https://user-images.githubusercontent.com/679486/92721939-1383d300-f324-11ea-8603-14e464fa156b.png)

This number will grow with the charge count.

Questions:
1. Is it important to have a different performance profile when paused?
2. Is it worth slowing down the rest of the sim by, say, 1%?  Perhaps more on other platforms.
3. If we want to pursue this, can it be optimized?  Perhaps we can speed up the matrix check (only check the scale/translation/rotation components)?
4. We would likely also need to round the opacity check for charge > 0 case.
5. Is this performance improvement good enough to reduce overall CPU and hence fan activity on MD's device?

Here is my patch from investigation:

<details>

```diff
Index: js/view/ChargeNode.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/view/ChargeNode.js	(revision 13d616c43bf5147e50490320335516278fd1980c)
+++ js/view/ChargeNode.js	(date 1599736411909)
@@ -111,8 +111,10 @@
       // Center
       NODE_MATRIX.multiplyMatrix( ARROW_OFFSET );
 
-      // Apply the transform
-      this.matrix = NODE_MATRIX;
+      // Apply the transform, but only if it is different enough to re-render.  (Small differences accrue when paused.)
+      if ( !this.matrix.equalsEpsilon( NODE_MATRIX, 1E-6 ) ) {
+        this.matrix = NODE_MATRIX;
+      }
 
       let opacity = Utils.linear( 0.015, CONVENTIONAL_CHARGE_THRESHOLD, 1, 0, Math.abs( charge.circuitElement.currentProperty.get() ) );
       opacity = Utils.clamp( opacity, 0, 1 );
@@ -129,8 +131,10 @@
       // Center the electrons
       NODE_MATRIX.multiplyMatrix( ELECTRON_OFFSET );
 
-      // Apply the transform
-      this.matrix = NODE_MATRIX;
+      // Apply the transform, but only if it is different enough to re-render.  (Small differences accrue when paused.)
+      if ( !this.matrix.equalsEpsilon( NODE_MATRIX, 1E-6 ) ) {
+        this.matrix = NODE_MATRIX;
+      }
     }
     this.updateVisible();
     this.outsideOfBlackBoxProperty.set( !charge.circuitElement.insideTrueBlackBoxProperty.get() );

```

</details>

@arouinfar or @ariel-phet can you please advise?

**Comment by ariel-phet** [comment]
@samreid before going further we need device info on MD's device. If it is a low performance device that would be good to know, and he is likely running quite an old version of windows etc (if past experience is an indicator).

If MD emailed you directly, lets gather this information before going any further.

**Comment by ariel-phet** [comment]
@samreid for instance, my fast home desktop running this sim barely notices a load (Quite new windows 10 PC running the sim in Chrome)

- less than 10% CPU usage on chrome,  ~15% total CPU when just having the sim run
- less than 15% CPU usage on chrome with quickly moving a slider back and forth for a lit lightbulb with electrons all, around 20% total CPU
- Chrome ~1% CPU when sim is closed, and just an idle webpage opened as a reference

**Comment by samreid** [comment]
I wrote:

> Thanks for reporting the CPU problem for Circuit Construction Kit.  I see from your report that you were using Chrome on Windows 10 on a Dell Latitude laptop.  Is this one of the student laptops, or your personal laptop?  Do you know about how old the laptop is?

MD replied:

> This is my personal laptop.  It is about a year old.

**Comment by ariel-phet** [comment]
@samreid I think we need one more piece of information. Is this CPU usage only with the CCK AC prototype or also with CCK DC for MD?

Correct me if I am wrong, but what you are suggesting above related to electron movement also exists in the DC version does it not? CCK DC has been published for quite some time now with no known performance complaints, so if this perceived performance issue exists in the DC version it does not seem worth pursuing to me.

**Comment by samreid** [comment]
MD reported this for the AC CCK, but it is most likely caused by a change in circuit-construction-kit-common which is used for DC and AC sims, which was introduced in April 2020 for https://github.com/phetsims/circuit-construction-kit-common/issues/572.

**Comment by ariel-phet** [comment]
@samreid @arouinfar let's plan to discuss at the next CCK design meeting

**Comment by samreid** [comment]
@ariel-phet suggests revising https://github.com/phetsims/circuit-construction-kit-common/issues/572 to have the physics engine run a very small dt if the user has added user input.  For instance, mouse down, mouse drag (while down), mouse up.  This would be on the entire display, including sliders etc.

**Comment by samreid** [comment]
I decided to work on this as part of #572, on hold until there is progress in that issue.

**Comment by samreid** [comment]
We addressed https://github.com/phetsims/circuit-construction-kit-common/issues/588 with a change-based approach rather than a small-dt-based approach, and that issue is ready for QA testing.  I measured performance while paused and CPU was minimal, most of the time idle:

![image](https://user-images.githubusercontent.com/679486/103059838-714eca80-4563-11eb-9a44-24d6f20f8451.png)

Therefore this issue can be closed.  I'll open a side issue for @kathy-phet's recommendation above.
